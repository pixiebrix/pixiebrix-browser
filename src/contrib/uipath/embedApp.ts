/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Renderer, UnknownObject } from "@/types";
import { isEmpty } from "lodash";
import { BlockArg, BlockOptions, Schema } from "@/core";
import { uuidv4 } from "@/types/helpers";
import browser, { Permissions } from "webextension-polyfill";
import { executeForNonce } from "@/background/executor";
import { unsafeAssumeValidArg } from "@/runtime/runtimeTypes";

export class UiPathAppRenderer extends Renderer {
  constructor() {
    super(
      "@pixiebrix/uipath/app",
      "UiPath App",
      "Render a UiPath App with optional parameters"
    );
  }

  permissions: Permissions.Permissions = {
    origins: ["https://cloud.uipath.com/*"],
  };

  inputSchema: Schema = {
    $schema: "https://json-schema.org/draft/2019-09/schema#",
    type: "object",
    properties: {
      url: {
        type: "string",
        description: "The full UiPath app URL",
      },
      inputs: {
        type: "object",
        description: "Optional inputs to pass to the app",
        additionalProperties: true,
      },
      title: {
        type: "string",
        description: "The title of the iframe",
        default: "UiPath App",
      },
      width: {
        type: "string",
        description: "The width of the iframe",
        default: "100%",
      },
      height: {
        type: "string",
        description: "The height of the iframe",
        default: "400",
      },
    },
    required: ["url"],
  };

  async render(
    {
      url,
      inputs: rawInputs = {},
      title = "UiPath App",
      height = 400,
      width = "100%",
    }: BlockArg,
    { logger }: BlockOptions
  ): Promise<string> {
    // https://transitory.technology/browser-extensions-and-csp-headers/
    const frameSrc = browser.runtime.getURL("frame.html");

    const nonce = uuidv4();

    const frameURL = new URL(frameSrc);
    frameURL.searchParams.set("url", url);
    frameURL.searchParams.set("nonce", nonce);

    const inputs = rawInputs as UnknownObject;

    if (!isEmpty(inputs)) {
      executeForNonce(
        nonce,
        "@pixiebrix/forms/set",
        unsafeAssumeValidArg({
          inputs: Object.entries(inputs).map(([key, value]) => ({
            selector: `[placeholder="in:${key}"]`,
            value,
          })),
        }),
        {
          isAvailable: {
            // UiPath apps lazy load the inputs, so make sure they've been rendered before trying
            // to set the values
            selectors: [
              isEmpty(inputs) ? ".root-container" : ".root-container input",
            ],
          },
          ctxt: {},
          messageContext: {
            blockId: this.id,
          },
        }
      ).catch((error) => {
        logger.error(error);
      });
    }

    return `<iframe src="${frameURL.toString()}" title="${title}" height="${height}" width="${width}" style="border:none;"></iframe>`;
  }
}
