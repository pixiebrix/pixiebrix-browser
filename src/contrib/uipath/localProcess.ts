/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { TransformerABC } from "@/types/bricks/transformerTypes";
import type { JobResult } from "@uipath/robot/dist/models";
import { validateRegistryId } from "@/types/helpers";
import { BusinessError } from "@/errors/businessErrors";
import { type Schema, type SchemaProperties } from "@/types/schemaTypes";
import { type BrickArgs, type BrickOptions } from "@/types/runtimeTypes";

export const UIPATH_ID = validateRegistryId("@pixiebrix/uipath/local-process");

const UIPATH_PROPERTIES: SchemaProperties = {
  releaseKey: {
    type: "string",
    description: "The local UiPath process id",
  },
  inputArguments: {
    type: "object",
    additionalProperties: true,
  },
};

export class RunLocalProcess extends TransformerABC {
  constructor() {
    super(
      UIPATH_ID,
      "Run local UiPath process",
      "Run a UiPath process using your local UiPath assistant"
    );
  }

  inputSchema: Schema = {
    $schema: "https://json-schema.org/draft/2019-09/schema#",
    type: "object",
    required: ["releaseKey"],
    properties: UIPATH_PROPERTIES,
  };

  async transform(
    { releaseKey, inputArguments = {} }: BrickArgs,
    { logger }: BrickOptions
  ): Promise<JobResult> {
    const { UiPathRobot } = await import(
      /* webpackChunkName: "uipath-robot" */ "@/contrib/uipath/UiPathRobot"
    );

    return Promise.race([
      // Throw error if Assistant is missing
      new Promise((resolve, reject) => {
        UiPathRobot.on("missing-components", () => {
          reject(new Error("UiPath Assistant not found. Is it installed?"));
        });
      }),

      // Run requested process
      (async () => {
        const processes = await UiPathRobot.init().getProcesses();

        const process = processes.find((x) => x.id === releaseKey);
        if (!process) {
          // `releaseKey`'s type is checked in the inputSchema
          logger.error(`Cannot find UiPath release: ${releaseKey as string}`);
          throw new BusinessError("Cannot find UiPath release");
        }

        console.debug("Running local UiPath process", { releaseKey });
        const result = await process.start(inputArguments);
        console.debug("Forwarding result from local UiPath process", {
          result,
          releaseKey,
        });
        return result;
      })(),
    ]);
  }
}
