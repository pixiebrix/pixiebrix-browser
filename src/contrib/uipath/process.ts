/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { proxyService } from "@/background/messenger/api";
import { TransformerABC } from "@/types/bricks/transformerTypes";
import { validateRegistryId } from "@/types/helpers";
import { BusinessError } from "@/errors/businessErrors";
import { type Schema, type SchemaProperties } from "@/types/schemaTypes";
import { type RegistryId } from "@/types/registryTypes";
import { type BrickArgs, type BrickOptions } from "@/types/runtimeTypes";
import { type SanitizedIntegrationConfig } from "@/types/integrationTypes";
import { type UnknownObject } from "@/types/objectTypes";
import { pollUntilTruthy } from "@/utils/promiseUtils";

export const UIPATH_SERVICE_IDS: RegistryId[] = [
  "uipath/cloud",
  "uipath/cloud-oauth",
  "uipath/orchestrator",
].map((x) => validateRegistryId(x));
export const UIPATH_ID = validateRegistryId("@pixiebrix/uipath/process");

const DEFAULT_MAX_WAIT_MILLIS = 30_000;
const POLL_MILLIS = 1000;

export const UIPATH_PROPERTIES: SchemaProperties = {
  uipath: {
    anyOf: [
      {
        $ref: "https://app.pixiebrix.com/schemas/services/uipath/cloud",
      },
      {
        $ref: "https://app.pixiebrix.com/schemas/services/uipath/orchestrator",
      },
      {
        $ref: "https://app.pixiebrix.com/schemas/services/uipath/cloud-oauth",
      },
    ],
  },
  releaseKey: {
    type: "string",
  },
  strategy: {
    type: "string",
    default: "JobsCount",
    enum: ["All", "Specific", "JobsCount"],
    description: "How the process should be run",
  },
  robotIds: {
    type: "array",
    items: {
      type: "number",
    },
  },
  jobsCount: {
    type: "number",
    description:
      "When using strategy JobsCount, the process will run x times where x is the value of the JobsCount field.",
  },
  awaitResult: {
    type: "boolean",
    default: false,
    description: "Wait for the process to complete and output the results.",
  },
  maxWaitMillis: {
    type: "number",
    default: DEFAULT_MAX_WAIT_MILLIS,
    description:
      "Maximum time (in milliseconds) to wait for the process to complete when awaiting result.",
  },
  inputArguments: {
    type: "object",
    additionalProperties: true,
  },
};

interface JobsResponse {
  "@odata.context": "https://cloud.uipath.com/odata/$metadata#Jobs";
  "@odata.count": number;
  value: Array<{
    Id: number;
    Key: string;
    State: "Successful" | "Pending" | "Faulted";
    OutputArguments: string;
    Info: string;
  }>;
}

export class RunProcess extends TransformerABC {
  constructor() {
    super(
      UIPATH_ID,
      "Run a UiPath process",
      "Run a UiPath process using UiPath Orchestrator"
    );
  }

  inputSchema: Schema = {
    $schema: "https://json-schema.org/draft/2019-09/schema#",
    type: "object",
    required: ["uipath", "releaseKey"],
    properties: UIPATH_PROPERTIES,
  };

  async transform(
    {
      uipath,
      releaseKey,
      strategy = "JobsCount",
      jobsCount = 0,
      robotIds = [],
      awaitResult = false,
      maxWaitMillis = DEFAULT_MAX_WAIT_MILLIS,
      inputArguments = {},
    }: BrickArgs<{
      uipath: SanitizedIntegrationConfig;
      releaseKey: string;
      strategy: string;
      jobsCount: number;
      robotIds: number[];
      awaitResult: boolean;
      maxWaitMillis: number;
      inputArguments: UnknownObject;
    }>,
    { logger }: BrickOptions
  ): Promise<unknown> {
    const responsePromise = proxyService<JobsResponse>(uipath, {
      url: "/odata/Jobs/UiPath.Server.Configuration.OData.StartJobs",
      method: "post",
      data: {
        startInfo: {
          ReleaseKey: releaseKey,
          Strategy: strategy,
          JobsCount: jobsCount,
          RobotIds: robotIds,
          Source: "Manual",
          InputArguments: JSON.stringify(inputArguments),
        },
      },
    });

    if (!awaitResult) {
      return {};
    }

    const { data: startData } = await responsePromise;

    if (startData.value.length > 1) {
      throw new BusinessError(
        "Awaiting response of multiple jobs not supported"
      );
    }

    const poll = async () => {
      const { data: resultData } = await proxyService<JobsResponse>(uipath, {
        url: `/odata/Jobs?$filter=Id eq ${startData.value[0].Id}`,
        method: "get",
      });

      if (resultData.value.length === 0) {
        logger.error(`UiPath job not found: ${startData.value[0].Id}`);
        throw new BusinessError("UiPath job not found");
      }

      if (resultData.value[0].State === "Successful") {
        return JSON.parse(resultData.value[0].OutputArguments);
      }

      if (resultData.value[0].State === "Faulted") {
        logger.error(`UiPath job failed: ${resultData.value[0].Info}`);
        throw new BusinessError("UiPath job failed");
      }
    };

    const result = await pollUntilTruthy(poll, {
      intervalMillis: POLL_MILLIS,
      maxWaitMillis,
    });

    if (result) {
      return result;
    }

    throw new BusinessError(
      `UiPath job did not finish in ${Math.round(maxWaitMillis / 1000)} seconds`
    );
  }
}
