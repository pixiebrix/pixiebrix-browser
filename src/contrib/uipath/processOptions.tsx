/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useEffect, useMemo } from "react";
import {
  BlockOptionProps,
  FieldRenderer,
  ServiceField,
} from "@/components/fields/blockOptions";
import { fromPairs, identity } from "lodash";
import {
  UIPATH_PROPERTIES,
  UIPATH_SERVICE_IDS,
} from "@/contrib/uipath/process";
import { Schema, SchemaProperties } from "@/core";
import { useField } from "formik";
import { useAsyncState } from "@/hooks/common";
import { proxyService } from "@/background/requests";
import { Button, Card, Form } from "react-bootstrap";
import { fieldLabel } from "@/components/fields/fieldUtils";
import Select from "react-select";
import { FieldProps } from "@/components/fields/propTypes";
import { parseAssemblyQualifiedName } from "csharp-helpers";
import { inputProperties } from "@/helpers";
import { useDependency } from "@/services/hooks";
import { BusinessError } from "@/errors";

interface Argument {
  name: string;
  type: string;
  required: boolean;
  hasDefault: boolean;
}

interface ODataResponseData<TValue> {
  "@odata.context": string;
  "@odata.count": number;
  value: TValue[];
}

interface Robot {
  MachineName: string;
  MachineId: number;
  Name: string;
  Username: string;
  Description: string;
  Type: string;
  Id: number;
}

interface Release {
  Key: string;
  ProcessKey: string;
  ProcessVersion: string;
  IsLatestVersion: boolean;
  Description: string;
  Name: string;
  Arguments: {
    // Serialized input dict
    Input: string | null;
    Output: string | null;
  };
}

export function useReleases(): {
  releases: Release[];
  isPending: boolean;
  error: unknown;
  hasConfig: boolean;
} {
  const { config, hasPermissions } = useDependency(UIPATH_SERVICE_IDS);

  const [releases, isPending, error] = useAsyncState(async () => {
    if (config && hasPermissions) {
      const response = await proxyService<ODataResponseData<Release>>(config, {
        url: "/odata/Releases",
        method: "get",
      });
      return response.data.value;
    }

    return null;
  }, [config, hasPermissions]);

  console.debug("releases", { releases, isPending, error });

  return { releases, isPending, error, hasConfig: config != null };
}

function useRobots(): { robots: Robot[]; isPending: boolean; error: unknown } {
  const { config, hasPermissions } = useDependency(UIPATH_SERVICE_IDS);
  const [robots, isPending, error] = useAsyncState(async () => {
    if (config && hasPermissions) {
      const response = await proxyService<ODataResponseData<Robot>>(config, {
        url: "/odata/Robots",
        method: "get",
      });
      return response.data.value;
    }

    return [];
  }, [config, hasPermissions]);

  return { robots, isPending, error };
}

const RobotsField: React.FunctionComponent<FieldProps<number[]>> = ({
  label,
  schema,
  ...props
}) => {
  const [{ value = [], ...field }, meta, helper] = useField<number[]>(props);

  const { robots, error } = useRobots();

  const options = useMemo(
    () => (robots ?? []).map((x) => ({ value: x.Id, label: x.Id })),
    [robots]
  );

  return (
    <Form.Group>
      <Form.Label>{label ?? fieldLabel(field.name)}</Form.Label>
      <Select
        isMulti
        options={options}
        value={options.filter((x) => value.includes(x.value))}
        onChange={(values) => {
          console.debug("Selected values", { values });
          helper.setValue((values as any)?.map((x: any) => x.value) ?? []);
        }}
      />
      {schema.description && (
        <Form.Text className="text-muted">The UIPath process to run</Form.Text>
      )}
      {error && (
        <span className="text-danger small">
          Error fetching robots: {error.toString()}
        </span>
      )}
      {meta.touched && meta.error && (
        <span className="text-danger small">{meta.error}</span>
      )}
    </Form.Group>
  );
};

export const ReleaseField: React.FunctionComponent<
  FieldProps<string> & { releases: Release[]; fetchError: unknown }
> = ({ label, schema, releases, fetchError, ...props }) => {
  const [{ value, ...field }, meta, helpers] = useField(props);

  const options = useMemo(
    () =>
      (releases ?? []).map((x) => ({
        value: x.Key,
        label: `${x.Name} - ${x.ProcessVersion}`,
        release: x,
      })),
    [releases]
  );

  return (
    <Form.Group>
      <Form.Label>{label ?? fieldLabel(field.name)}</Form.Label>
      <Select
        options={options}
        value={options.find((x) => x.value === value)}
        onChange={(option) => helpers.setValue((option as any)?.value)}
      />
      {schema.description && (
        <Form.Text className="text-muted">The UIPath process to run</Form.Text>
      )}
      {fetchError && (
        <span className="text-danger small">
          Error fetching releases: {fetchError.toString()}
        </span>
      )}
      {meta.touched && meta.error && (
        <span className="text-danger small">{meta.error}</span>
      )}
    </Form.Group>
  );
};

function toType(type: string) {
  const { namespace, typeName } = parseAssemblyQualifiedName(type);
  // https://docs.microsoft.com/en-us/dotnet/api/system.valuetype?view=net-5.0
  if (namespace === "System" && typeName === "String") {
    return "string";
  }

  if (namespace === "System" && typeName === "Boolean") {
    return "boolean";
  }

  if (
    namespace === "System" &&
    ["Int64", "Int32", "Int16", "UInt64", "UInt32", "UInt16"].includes(typeName)
  ) {
    return "integer";
  }

  if (namespace === "System" && ["Decimal", "Double"].includes(typeName)) {
    return "number";
  }

  throw new BusinessError(`Unsupported input type: ${type}`);
}

export function releaseSchema(release: Release): Schema {
  if (!release.Arguments.Input) {
    return {};
  }

  const inputs = JSON.parse(release.Arguments.Input) as Argument[];

  const properties = fromPairs(
    inputs.map((input) => [input.name, { type: toType(input.type) }])
  ) as SchemaProperties;

  return {
    properties,
    required: inputs.filter((input) => input.required).map((x) => x.name),
  };
}

export const InputArgumentsField: React.FunctionComponent<
  FieldProps<object>
> = ({ name, schema, label }) => (
  <Form.Group>
    <Form.Label>inputArguments</Form.Label>
    <Card>
      <Card.Header>{label}</Card.Header>
      <Card.Body>
        {schema &&
          Object.entries(inputProperties(schema)).map(([prop, fieldSchema]) => {
            if (typeof fieldSchema === "boolean") {
              throw new TypeError("Expected schema for input property type");
            }

            return (
              <FieldRenderer
                key={prop}
                name={`${name}.${prop}`}
                schema={schema}
              />
            );
          })}
      </Card.Body>
    </Card>
  </Form.Group>
);

const ProcessOptions: React.FunctionComponent<BlockOptionProps> = ({
  name,
  configKey,
  showOutputKey,
}) => {
  const basePath = [name, configKey].filter(identity).join(".");

  const { hasPermissions, requestPermissions } = useDependency(
    UIPATH_SERVICE_IDS
  );

  const [{ value: releaseKey }] = useField<string>(`${basePath}.releaseKey`);
  const [{ value: strategy }, , strategyHelpers] = useField<string>(
    `${basePath}.strategy`
  );
  const [{ value: jobsCount }, , jobsCountHelpers] = useField<number>(
    `${basePath}.jobsCount`
  );

  const { releases, error: releasesError } = useReleases();

  useEffect(() => {
    if (!strategy) {
      strategyHelpers.setValue((UIPATH_PROPERTIES["strategy"] as any).default);
    } else if (strategy === "JobsCount" && jobsCount == null) {
      jobsCountHelpers.setValue(1);
    }
  }, [strategy, jobsCount]);

  const [release, schema] = useMemo(() => {
    const release = releases?.find((x) => x.Key === releaseKey);
    const schema = release ? releaseSchema(release) : null;
    return [release, schema];
  }, [releases, releaseKey]);

  if (!hasPermissions) {
    return (
      <div className="my-2">
        <p>
          You must grant permissions for you browser to send information to
          UiPath.
        </p>
        <Button onClick={requestPermissions}>Grant Permissions</Button>
      </div>
    );
  }

  return (
    <div>
      <ServiceField
        key="uipath"
        name={`${basePath}.uipath`}
        schema={UIPATH_PROPERTIES["uipath"] as Schema}
      />
      <ReleaseField
        label="release"
        name={`${basePath}.releaseKey`}
        schema={UIPATH_PROPERTIES["releaseKey"] as Schema}
        releases={releases}
        fetchError={releasesError}
      />
      <FieldRenderer
        name={`${basePath}.strategy`}
        schema={UIPATH_PROPERTIES["strategy"] as Schema}
      />

      {strategy === "Specific" && (
        <RobotsField
          name={`${basePath}.robotIds`}
          schema={UIPATH_PROPERTIES["robotIds"] as Schema}
        />
      )}
      {strategy === "JobsCount" && (
        <FieldRenderer
          name={`${basePath}.jobsCount`}
          schema={UIPATH_PROPERTIES["jobsCount"] as Schema}
        />
      )}
      <FieldRenderer
        name={`${basePath}.awaitResult`}
        schema={UIPATH_PROPERTIES["awaitResult"] as Schema}
      />
      {releaseKey && release && (
        <InputArgumentsField
          label={release.Name}
          schema={schema}
          name={[name, configKey, "inputArguments"].filter(identity).join(".")}
        />
      )}

      {showOutputKey && (
        <FieldRenderer
          name={`${name}.outputKey`}
          label="Output Variable"
          schema={{
            type: "string",
            description: "A name to refer to this brick in subsequent bricks",
          }}
        />
      )}
    </div>
  );
};

export default ProcessOptions;
