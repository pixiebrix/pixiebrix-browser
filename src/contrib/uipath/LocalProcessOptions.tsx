/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { useMemo, useState } from "react";
import { partial } from "lodash";
import { UIPATH_PROPERTIES as REMOTE_UIPATH_PROPERTIES } from "@/contrib/uipath/process";
import { Schema } from "@/core";
import { useAsyncEffect } from "use-async-effect";
import ChildObjectField from "@/components/fields/schemaFields/ChildObjectField";
import { BlockOptionProps } from "@/components/fields/schemaFields/genericOptionsFactory";
import { useSelectedRelease } from "@/contrib/uipath/uipathHooks";
import { joinName } from "@/utils";
import RequireServiceConfig from "@/contrib/RequireServiceConfig";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import RemoteSelectWidget from "@/components/form/widgets/RemoteSelectWidget";
import { thisTab } from "@/devTools/utils";
import { getProcesses, initRobot } from "@/contentScript/messenger/api";
import { isDevToolsPage } from "webext-detect-page";

function useLocalRobot() {
  const [robotAvailable, setRobotAvailable] = useState(false);
  const [consentCode, setConsentCode] = useState(null);
  const [initError, setInitError] = useState(null);

  useAsyncEffect(async () => {
    if (!isDevToolsPage()) {
      setInitError(
        new Error("UiPath Assistant can only be configured from a page context")
      );
      return;
    }

    try {
      const { available, consentCode } = await initRobot(thisTab);
      setConsentCode(consentCode);
      setRobotAvailable(available);
    } catch (error) {
      setInitError(error);
    }
  }, [setConsentCode, setRobotAvailable, setInitError]);

  return {
    robotAvailable,
    consentCode,
    initError,
  };
}

const LocalProcessOptions: React.FunctionComponent<BlockOptionProps> = ({
  name,
  configKey,
}) => {
  const configName = partial(joinName, name, configKey);

  const { robotAvailable, consentCode } = useLocalRobot();
  const { selectedRelease } = useSelectedRelease(configName("releaseKey"));

  const processesPromise = useMemo(async () => {
    if (robotAvailable) {
      return getProcesses(thisTab);
    }

    return [];
  }, [robotAvailable]);

  if (!robotAvailable) {
    return (
      <div>
        <span className="text-danger">
          UiPath Assistant not found. Don&apos;t have the UiPath Assistant?{" "}
          <a
            href="https://robotjs.uipath.com/download"
            target="_blank"
            rel="noopener noreferrer"
          >
            Get it now.
          </a>
        </span>
      </div>
    );
  }

  return (
    <div>
      {consentCode && (
        <span className="text-info">
          UiPath Assistant consent code: {consentCode}
        </span>
      )}

      <RequireServiceConfig
        // FIXME: this service use is options-only. As-is this will create an integration entry in the background. We
        //  need to support 1) making RemoteServiceConfig optional, and 2) not storing the state in Formik
        serviceSchema={REMOTE_UIPATH_PROPERTIES.uipath as Schema}
        serviceFieldName={configName("service")}
      >
        {() => (
          <>
            <ConnectedFieldTemplate
              label="Process"
              description="Select a local process"
              name={configName("releaseKey")}
              as={RemoteSelectWidget}
              blankValue={null}
              optionsFactory={processesPromise}
            />

            <ChildObjectField
              heading={selectedRelease?.release?.Name ?? "Input Arguments"}
              schema={selectedRelease?.schema}
              name={configName("inputArguments")}
            />
          </>
        )}
      </RequireServiceConfig>
    </div>
  );
};

export default LocalProcessOptions;
