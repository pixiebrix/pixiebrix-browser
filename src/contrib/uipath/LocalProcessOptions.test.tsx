/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { render } from "@testing-library/react";
import LocalProcessOptions from "@/contrib/uipath/LocalProcessOptions";
import * as contentScriptApi from "@/contentScript/messenger/api";
import { Formik } from "formik";
import { menuItemFormStateFactory } from "@/testUtils/factories";
import { UIPATH_ID } from "@/contrib/uipath/localProcess";
import { waitForEffect } from "@/testUtils/testHelpers";
import { uuidv4, validateRegistryId } from "@/types/helpers";
import { type OutputKey, type SanitizedServiceConfiguration } from "@/core";
import * as auth from "@/hooks/auth";
import * as dependencyHooks from "@/services/useDependency";
import { type Service } from "@/types";
import { type FormState } from "@/pageEditor/extensionPoints/formStateTypes";

jest.mock("webext-detect-page", () => ({
  isDevToolsPage: () => true,
  isExtensionContext: () => true,
  isBackground: () => false,
  isContentScript: () => false,
}));

jest.mock("@/services/useDependency", () =>
  jest.fn().mockReturnValue({
    // Pass minimal arguments
    config: undefined,
    service: undefined,
    hasPermissions: true,
    requestPermissions: jest.fn(),
  })
);
jest.mock("@/hooks/auth");
jest.mock("@/contrib/uipath/uipathHooks");
jest.mock("@/hooks/auth");
jest.mock("@/contentScript/messenger/api");
jest.mock("@/background/messenger/api");
jest.mock("@/contrib/uipath/uipathHooks", () => {
  const mock = jest.requireActual("@/contrib/uipath/uipathHooks");
  return {
    __esModule: true,
    ...mock,
    useSelectedRelease: jest.fn().mockResolvedValue({
      selectedRelease: null,
      releasePromise: Promise.resolve([]),
    }),
  };
});
jest.mock("@/components/form/widgets/RemoteSelectWidget", () => {
  const mock = jest.requireActual(
    "@/components/form/widgets/RemoteSelectWidget"
  );
  return {
    __esModule: true,
    ...mock,
    useOptionsResolver: jest.fn().mockReturnValue([[], false, undefined]),
  };
});

const serviceId = validateRegistryId("@uipath/cloud");

function makeBaseState() {
  const baseFormState = menuItemFormStateFactory();
  baseFormState.services = [
    { id: serviceId, outputKey: "uipath" as OutputKey },
  ];
  baseFormState.extension.blockPipeline = [
    {
      id: UIPATH_ID,
      config: {
        service: null,
        releaseKey: null,
        inputArguments: {},
      },
    },
  ];
  return baseFormState;
}

function renderOptions(formState: FormState = makeBaseState()) {
  return render(
    <Formik onSubmit={jest.fn()} initialValues={formState}>
      <LocalProcessOptions
        name="extension.blockPipeline.0"
        configKey="config"
      />
    </Formik>
  );
}

describe("UiPath LocalProcess Options", () => {
  test("Can render options", async () => {
    (
      contentScriptApi.getProcesses as jest.MockedFunction<
        typeof contentScriptApi.getProcesses
      >
    ).mockResolvedValue([]);
    (
      contentScriptApi.initRobot as jest.MockedFunction<
        typeof contentScriptApi.initRobot
      >
    ).mockResolvedValue({
      available: false,
      consentCode: null,
      missingComponents: false,
    });

    const rendered = renderOptions();

    await waitForEffect();

    expect(rendered.queryByText("UiPath Assistant not found")).toBeNull();
    expect(rendered.container).toMatchSnapshot();
  });

  test("Can render consent code and service selector", async () => {
    (
      auth.useAuthOptions as jest.MockedFunction<typeof auth.useAuthOptions>
    ).mockReturnValue([[], jest.fn()]);
    (
      contentScriptApi.getProcesses as jest.MockedFunction<
        typeof contentScriptApi.getProcesses
      >
    ).mockResolvedValue([]);
    (
      contentScriptApi.initRobot as jest.MockedFunction<
        typeof contentScriptApi.initRobot
      >
    ).mockResolvedValue({
      available: true,
      consentCode: "abc123",
      missingComponents: false,
    });

    const rendered = renderOptions();

    await waitForEffect();

    expect(
      rendered.queryByText("UiPath Assistant consent code: abc123")
    ).not.toBeNull();
    expect(rendered.queryByLabelText("Integration")).not.toBeNull();
  });

  test("Can render arguments", async () => {
    const config = uuidv4();

    (
      dependencyHooks.default as jest.MockedFunction<
        typeof dependencyHooks.default
      >
    ).mockReturnValue({
      // Pass minimal arguments
      config: {
        id: config,
        serviceId,
      } as unknown as SanitizedServiceConfiguration,
      service: { id: serviceId } as unknown as Service,
      hasPermissions: true,
      requestPermissions: jest.fn(),
    });
    (
      auth.useAuthOptions as jest.MockedFunction<typeof auth.useAuthOptions>
    ).mockReturnValue([
      [{ label: "Test Auth", value: config, serviceId, local: true }],
      jest.fn(),
    ]);
    (
      contentScriptApi.getProcesses as jest.MockedFunction<
        typeof contentScriptApi.getProcesses
      >
    ).mockResolvedValue([]);
    (
      contentScriptApi.initRobot as jest.MockedFunction<
        typeof contentScriptApi.initRobot
      >
    ).mockResolvedValue({
      available: true,
      consentCode: null,
      missingComponents: false,
    });

    const formState = makeBaseState();
    formState.services[0].config = config;
    formState.extension.blockPipeline[0].config.service = "@uipath";

    const rendered = renderOptions();

    await waitForEffect();

    expect(rendered.queryByText("UiPath Assistant consent code")).toBeNull();
    expect(rendered.queryByLabelText("Integration")).not.toBeNull();
    expect(rendered.queryByLabelText("Process")).not.toBeNull();
    expect(rendered.queryByText("Input Arguments")).not.toBeNull();
    expect(rendered.queryByText("Add Property")).not.toBeNull();
  });
});
