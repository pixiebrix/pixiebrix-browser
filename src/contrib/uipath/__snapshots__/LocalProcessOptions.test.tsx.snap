// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UiPath LocalProcess Options Can render options 1`] = `
<div>
  <div>
    <span
      class="text-danger"
    >
      UiPath Assistant not found. Don't have the UiPath Assistant?
       
      <a
        href="https://robotjs.uipath.com/download"
        rel="noopener noreferrer"
        target="_blank"
      >
        Get it now.
      </a>
    </span>
  </div>
</div>
`;
